### ISSUE TITLE
Feature request: Keep screen active (Android)

### ISSUE THREAD

**Issue by Dizzap** [martinstoeckli/SilentNotes#149]
Hi,
fine feature I thought of randomly:
Keep the phone screen from automatically turning off while in the shopping mode. Probably with a global setting for users to turn this off/on.
Basic behaviour: `open a note > click shopping mode > screen stays turned on until turned off manually or the shopping mode gets turned off or the app gets minimized.`
I found myself in the need of this feature while shopping, since my screen is set to turn off after like 15 seconds.
No clue how difficult it is to implement. Would this be something of interest?

Cheers,
Jakub

**Comment by martinstoeckli** [comment]
Hi Jakub

A timeout of 15 sec would highly stress me, you can make life so much more relaxing when you set it to 5 min instead, try it out (it's the first setting I ever make on new phones) 😎 But the idea by itself is good, I'm not familiar with the API but it could become a handy feature. My thoughts about it:

1. I try to keep the required Android privileges to a minimum, so I would be glad if it does not require too many exotic privileges.
2. Maybe it would be better, if one could define a longer timeout instead of turning it off completely, but I don't know what's possible.
3. Perhaps the timeout could affect all notes, not only the ones opened in the shopping mode. I myself am using this mode also to "protect" notes, and a cooking recipe would also profit from the longer timeout.

Thanks a lot:
Martin

**Comment by Dizzap** [comment]
I actually can't change the timing, because I'm using a battery saver mode constantly, so my phone doesn't die in like 3 hours 😅
Anyway...
 - How would that "longer timeout" look like? Like a global setting or for specific note? Maybe even have a sliding scale going from 1 minute to never? 
 - Maybe we could do like "enabled for all, only in shopping mode, disabled" setting, so the user can decide? 
 - Maybe we could add a button to "lock" the screen from turning off on the note, same as the pin/shopping? It would let you change the behaviour ad hoc instead of having a general setting for every note. But it would also introduce another button to the UI, and we might as well use the shopping mode at that point...

**Comment by martinstoeckli** [comment]
The more I think about your idea, the more I like it 💛 The third point you mentioned triggered another idea, it would make operation really simple, so what you think:

- We add a menu entry on the note/checklist page called "Keep open for 20 min".
- I the settings this duration is configurable, some fix steps like 5min, 10min, 15min may be enough.

This way we wouldn't have a state which must be resetted, the user clicks the menu and can forget about it. The app stays open for a certain time and switches back to normal closing mode afterwards. There would be no further decicions to make when switching to this mode, no options for which notes, no duration for this time, just one fast button click.

**Comment by Dizzap** [comment]
Sounds cool. I'll look up how to control the screen timeout, then we can decide on the final form.

**Comment by Dizzap** [comment]
**Option 1 - Using Xamarin.Essentials**
- requires nuget package `Xamarin.Essentials` to be added
- requires `WAKE_LOCK` permission to be added

Other than that, it's seems fairly simple:
```
Xamarin.Essentials.Platform.Init(this, bundle); //otherwise it threw NullReferenceException
DeviceDisplay.KeepScreenOn = true;
```
So we could keep a timer to measure the timeout after pressing the lock button and then set it to false again. From my (very quick) testing, this works only if the app is actively on screen (not minimized), which is perfect.

Source: [https://andyp.dev/posts/xamarin-forms-how-to-prevent-android-from-auto-locking](https://andyp.dev/posts/xamarin-forms-how-to-prevent-android-from-auto-locking)

I'll keep looking for other solutions.

**Comment by martinstoeckli** [comment]
Interesting, thanks for the research. I had a quick look at the Xamarin.Essentials code and it seems that it boils down to setting the flag like this:

  https://developer.android.com/training/scheduling/wakelock

There the docu states that setting the flag doesn't even need the permission, which is great news. At the moment I'm a bit reluctant with including the Xamarin library, because sooner or later I would like to switch to its successor .Net 6/Maui, and because it seems to be just a one-liner.

**Comment by Dizzap** [comment]
Well, it doesn't look promising...
The only possible way I found is to edit the `SilentNotes.Andriod.Resources.Main.axml` and add `android:keepScreenOn="true"` to the WebView. This doesn't need any permissions or an additional library, but I doubt it can be manipulated afterwards. 
Do you have any idea if it's possible to do something like binding it to a global variable and then editing it programatically?

**Comment by Dizzap** [comment]
Lol, just as I wrote that, I found a post with a possible solution:
```
Window.AddFlags(WindowManagerFlags.KeepScreenOn);
Window.ClearFlags(WindowManagerFlags.KeepScreenOn);
```
The post link: https://stackoverflow.com/a/45450683

**Comment by martinstoeckli** [comment]
Yes, for everything you can define in XAML there should be a way to do it in code 👍

**Comment by martinstoeckli** [comment]
Hi Jakup

Did a pull request and added you as reviewer, because this was actually your idea. I saw that you removed the assignment, but wanted it to be part of the upcoming version 7.0, so this is a possible implementation. If you want, have a look at the code (but you don't have to of course).

**Comment by Dizzap** [comment]
Hi Martin,
Amazing! I removed myself because I had 0 time for the past months and didn't want to block you or anyone else from implementing this😅 (and the other features as well). I'll have a look at it in the evening 🙂

**Comment by martinstoeckli** [comment]
I just merged the pull request, so I do not create conflicts with further implementations, but the branch still exists and your feedback is welcome.

**Comment by Dizzap** [comment]
Had a look at the code and tried it on android and it looks really good! 🙂

**Comment by martinstoeckli** [comment]
Just published version [7.0.0](https://github.com/martinstoeckli/SilentNotes/releases/tag/v7.0.0) which implements this feature.
